File: src/vpn_automation/redaction.py
import re
from typing import Any


SECRET_DEPLOYMENT_KEYS = {
    "subscription_url",
    "verify_subscription_url",
    "secret_query",
    "share_project_sub_value",
    "pages_secret_admin",
}

SECRET_TEXT_PATTERNS = (
    (re.compile(r"(token=)[^&\s\"']+"), r"\1<redacted>"),
    (re.compile(r"(serect_key=)[^&\s\"']+"), r"\1<redacted>"),
    (re.compile(r"(secret_key=)[^&\s\"']+"), r"\1<redacted>"),
    (re.compile(r"(api[_-]?token=)[^&\s\"']+", re.IGNORECASE), r"\1<redacted>"),
    (re.compile(r"vmess://[A-Za-z0-9_\-+/=]+"), "vmess://<redacted>"),
)


def redact_text(value: str) -> str:
    redacted = value
    for pattern, replacement in SECRET_TEXT_PATTERNS:
        redacted = pattern.sub(replacement, redacted)
    return redacted


def safe_deployment(deployment: dict[str, Any]) -> dict[str, Any]:
    safe: dict[str, Any] = {}
    for key, value in deployment.items():
        if key in SECRET_DEPLOYMENT_KEYS:
            safe[key] = "set" if value else ""
        elif isinstance(value, str):
            safe[key] = redact_text(value)
        elif isinstance(value, (int, float, bool)) or value is None:
            safe[key] = value
        elif isinstance(value, list):
            safe[key] = [_redact_nested(item) for item in value]
        elif isinstance(value, dict):
            safe[key] = {str(nested_key): _redact_nested(nested_value) for nested_key, nested_value in value.items()}
        else:
            safe[key] = str(type(value).__name__)
    return safe


def _redact_nested(value: Any) -> Any:
    if isinstance(value, str):
        return redact_text(value)
    if isinstance(value, (int, float, bool)) or value is None:
        return value
    if isinstance(value, list):
        return [_redact_nested(item) for item in value]
    if isinstance(value, dict):
        return {str(key): _redact_nested(item) for key, item in value.items()}
    return str(type(value).__name__)

File: npm/autovpn-cli/src/runtime/redaction.ts
const SECRET_DEPLOYMENT_KEYS = new Set([
  'subscription_url',
  'verify_subscription_url',
  'secret_query',
  'share_project_sub_value',
  'pages_secret_admin'
]);

const SECRET_FIELD_NAMES = [
  'token',
  'serect_key',
  'secret_key',
  'api_token',
  'api-token',
  'cloudflare_api_token',
  'subscription_url',
  'verify_subscription_url',
  'secret_query',
  'share_project_sub_value',
  'pages_secret_admin'
];

const SECRET_FIELD_PATTERN = new RegExp(
  `(["']?)\\b(${SECRET_FIELD_NAMES.map((key) => key.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&')).join('|')})\\b\\1(\\s*[:=]\\s*)(["']?)([^\\s"',}&]+)(\\4)`,
  'gi'
);

const SECRET_TEXT_PATTERNS: Array<[RegExp, string | ((...args: string[]) => string)]> = [
  [SECRET_FIELD_PATTERN, (_match, keyQuote, key, separator, valueQuote, _value, closingQuote) => `${keyQuote}${key}${keyQuote}${separator}${valueQuote}<redacted>${closingQuote}`],
  [/(Bearer\s+)[A-Za-z0-9._~+/\-=]+/gi, '$1<redacted>'],
  [/vmess:\/\/[A-Za-z0-9_\-+/=]+/g, 'vmess://<redacted>']
];

export function redactText(value: string): string {
  return SECRET_TEXT_PATTERNS.reduce((result, [pattern, replacement]) => result.replace(pattern, replacement as string), value);
}

function redactNested(value: unknown): unknown {
  if (typeof value === 'string') {
    return redactText(value);
  }
  if (value === null || ['number', 'boolean'].includes(typeof value)) {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map(redactNested);
  }
  if (typeof value === 'object') {
    return Object.fromEntries(Object.entries(value as Record<string, unknown>).map(([key, item]) => [key, redactNested(item)]));
  }
  return typeof value;
}

export function safeDeployment(deployment: Record<string, unknown>): Record<string, unknown> {
  const safe: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(deployment)) {
    if (SECRET_DEPLOYMENT_KEYS.has(key)) {
      safe[key] = value ? 'set' : '';
    } else {
      safe[key] = redactNested(value);
    }
  }
  return safe;
}
